framework/source/class/qx/io/DynamicScriptLoader.js

```diff
@@ -0,0 +1,245 @@
+ *        var data = e.getData();
+ *        if (data.script == lastScript){
+ *          console.log('Last Script Loaded');
+ *          dynLoader.removeListenerById(loadId);
+ *        }
+ *      }, this);
+ *    }
+ *    else {
+ *      console.log('Scripts have already been loaded');
+ *    }
+ * </code>
+ */
+qx.Class.define("qx.io.DynamicScriptLoader", {
+  extend: qx.core.Object,
+
+  type: "singleton",
+
+  /**
+   * Initialize the state hashes if not yet done
+   */
+
+  construct: function () {
+    this.base(arguments);
+    this.__ADDED = {};
+    this.__LOADED = {};
+    this.__QUEUE = [];
+    this.__LOADING = false;
+  },
+
+  /*
+  *****************************************************************************
+     EVENTS
+  *****************************************************************************
+  */
+
+  events: {
+    /**
+     * fired when a specific script is loaded successfully
+     */
+    loaded: 'qx.event.type.Data',
+
+    /**
+     * fired when a specific script fails loading
+     */
+    failed: 'qx.event.type.Data',
+
+    /**
+     * fired when all given scripts are loaded, each time loadScriptsDynamic is called
+     */
+    ready: 'qx.event.type.Event'
+  },
+
+
+  /*
+  *****************************************************************************
+     MEMBERS
+  *****************************************************************************
+  */
+
+
+  members: {
+
+    __ADDED: null,
+    __LOADED: null,
+    __QUEUE: null,
+    __LOADING: null,
+
+
+    /**
+     * Load scripts dynamically, typically used in the instance constructor
+     *
+     * @param codeArr {Array} an array with the uri names of the scripts
+     * @param callback {Function?undefined} a callback function to call when all the scripts in the array have been loaded
+     * @param context {Object?undefined} a contex for calling the callback
+     *
+     * @return {String|null} name of the last script to be loaded
+     *     Returns the name of the last script you can expect to be loaded if all scripts are already loaded null will be returned.
+     *
+     */
+    load: function(codeArr, callback, context) {
+      var lastScript = null;
+      codeArr.forEach(function(script) {
+        if (this.__ADDED[script] !== true) {
+          this.__QUEUE.push(script);
+          lastScript = script;
+          this.__ADDED[script] = true;
+        }
+      }, this);
+      if (callback) {
+        var cb = new qx.util.DeferredCall(callback, context);
+        if (lastScript !== null) {
+          var id = this.addListener('loaded', function(e) {
+            var data = e.getData();
+            // this.debug("Loading " + data.script + " completed");
+            if (data.script == lastScript) {
+              this.removeListenerById(id);
+              cb.schedule();
+            }
+          }, this);
+        }
+        else {
+          cb.schedule();
+        }
+      }
+      this.__loadScripts();
```

Comment:
__loadScripts can't be a member function. It must be a function within the load() function, so that events are on a per-load() basis, not on a global basis.